File: src/II039.java
import java.util.Stack;

public class II039 {

}

class Solution {
    public int largestRectangleArea(int[] heights) {
        int ret = 0, n = heights.length;
        Stack<Integer> st = new Stack<>();
        int[] left = new int[n];
        for (int i = 0; i < n; ++i) {
            int height = heights[i];
            while (!st.empty() && heights[st.peek()] >= height) {
                st.pop();
            }
            left[i] = st.empty() ? -1 : st.peek();
            st.push(i);
        }
        st.clear();
        for (int i = n - 1; i >= 0; --i) {
            int height = heights[i];
            while (!st.empty() && heights[st.peek()] >= height) {
                st.pop();
            }
            int right = st.empty() ? n : st.peek();
            st.push(i);
            ret = Math.max(ret, (right - left[i] - 1) * heights[i]);
        }
        return ret;
    }
}

File: src/II039.ts
// function largestRectangleArea(heights: number[]): number {
//     heights.push(0);
//     let st: number[] = [-1], len = heights.length;
//     let ret = 0;
//     for (let i = 0; i < len; ++i) {
//         while (heights.at(st.at(-1) as number) as number > heights[i]) {
//             let height = heights[st.pop() as number];
//             let width = i - (st.at(-1) as number) - 1;
//             ret = Math.max(ret, height * width);
//         }
//         st.push(i);
//     }
//     return ret;
// };
